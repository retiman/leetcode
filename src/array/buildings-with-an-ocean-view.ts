// DIFFICULTY: MEDIUM
//
// There are n buildings in a line. You are given an integer array heights of size n that represents the heights of the
// buildings in the line.
//
// The ocean is to the right of the buildings. A building has an ocean view if the building can see the ocean without
// obstructions. Formally, a building has an ocean view if all the buildings to its right have a smaller height.
//
// Return a list of indices (0-indexed) of buildings that have an ocean view, sorted in increasing order.
//
// See {@link https://leetcode.com/problems/buildings-with-an-ocean-view/}
export { findBuildings };

// SOLUTION:
//
// This is easier to do iterating from right to left (since the ocean is to the right).  We can keep track of the
// tallest building we've seen so far, and if we encounter a building that is taller, we can add it to the list with
// ocean views.
//
// COMPLEXITY:
//
// Time complexity is O(n) because we are iterating through the list of buildings once.  Space complexity is O(n)
// because we are storing a result array.
function findBuildings(heights: number[]): number[] {
  const result: number[] = [];
  let tallest = -Infinity;

  for (let i = heights.length - 1; i >= 0; i--) {
    const height = heights[i];
    if (height > tallest) {
      tallest = height;
      result.push(i);
    }
  }

  // Problem says the buildings can't be in any order; they have to be in increasing order (of index, not height).
  return result.reverse();
}
