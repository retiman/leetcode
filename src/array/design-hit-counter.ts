// DIFFICULTY: Medium
//
// Design a hit counter which counts the number of hits received in the past 5 minutes (i.e., the past 300 seconds).
//
// Your system should accept a timestamp parameter (in seconds granularity), and you may assume that calls are being
// made to the system in chronological order (i.e., timestamp is monotonically increasing). Several hits may arrive
// roughly at the same time.
//
// See {@link https://leetcode.com/problems/design-hit-counter/}
export { HitCounter };

// SOLUTION:
//
// To do this efficiently we'll have to use a circular array buffer.  This is the same technique used by time series
// databases.
//
// COMPLEXITY:
//
// Both methods run in O(1) time since we fix the array size at 300.
interface HitEvent {
  timestamp: number;
  hits: number;
}

class HitCounter {
  private readonly events: HitEvent[];

  constructor() {
    // Our granularity is seconds, so in theory, we only need to store only the last 300 seconds.  Keep track of both
    // hits by timestamp.
    this.events = [];

    // Don't use Array.fill() here; you'll fill the array with 300 copies of the same object.
    for (let i = 0; i < 300; i++) {
      this.events[i] = {
        timestamp: 0,
        hits: 0
      };
    }
  }

  hit(timestamp: number): void {
    // Find the event in our circular buffer that would correspond to this timestamp.
    const i = timestamp % 300;
    const event = this.events[i];

    // If we've already recorded hits at this timestamp, increment it.
    if (event.timestamp === timestamp) {
      event.hits++;
      return;
    }

    // Otherwise, we haven't, so set the hits to 1.
    event.timestamp = timestamp;
    event.hits = 1;
  }

  getHits(timestamp: number): number {
    let count = 0;

    for (const event of this.events) {
      if (timestamp - event.timestamp < 300) {
        count += event.hits;
      }
    }

    return count;
  }
}
