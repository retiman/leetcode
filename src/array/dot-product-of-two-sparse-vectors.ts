// Given two sparse vectors, compute their dot product.
//
// Implement class SparseVector:
//
// - SparseVector(nums) Initializes the object with the vector nums
// - dotProduct(vec) Compute the dot product between the instance of SparseVector and vec
//
// A sparse vector is a vector that has mostly zero values, you should store the sparse vector efficiently and compute
// the dot product between two SparseVector.
//
// Follow up: What if only one of the vectors is sparse?
//
// See {@link https://leetcode.com/problems/dot-product-of-two-sparse-vectors/}
export { SparseVector };

// SOLUTION:
//
// If the vector is sparse, just store the non-zero values in a map of index to value.  That will compress for storage;
// to compute the dot product you can either decompress and then do the dot product, or you can do the multiplication
// over the non-zero values directly.
//
// Doing it without decompression is more efficient, and we should start with the smaller vector to minimize the number
// of operations.
class SparseVector {
  private readonly map: Map<number, number>;

  constructor(nums: number[]) {
    // This does the compression.
    this.map = new Map();
    for (let i = 0; i < nums.length; i++) {
      if (nums[i] !== 0) {
        this.map.set(i, nums[i]);
      }
    }
  }

  dotProduct(vec: SparseVector): number {
    // Get the vectors in order of size; we'll iterate over the smaller vector.
    const [a, b] = this.sorted(this.map, vec.map);

    // Iterate over the smaller vector and multiply the values, only if the larger vector has the same index.
    let result = 0;
    for (const [i, ai] of a) {
      if (!b.has(i)) {
        continue;
      }

      const bi = b.get(i)!;
      result += ai * bi;
    }

    return result;
  }

  // If one vector is dense, don't bother compressing it.  Just compute directly.
  dotProductDense(b: number[]) {
    let result = 0;
    for (const [i, ai] of this.map) {
      const bi = b[i];
      result += ai * bi;
    }

    return result;
  }

  private sorted(a: Map<number, number>, b: Map<number, number>): [Map<number, number>, Map<number, number>] {
    return a.size < b.size ? [a, b] : [b, a];
  }
}
