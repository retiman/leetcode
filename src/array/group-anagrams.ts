// DIFFICULTY: Medium
//
// Given an array of strings `strs`, group the anagrams together. You can return the answer in any order.
//
// An anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all
// the original letters exactly once.
//
// See {@link https://leetcode.com/problems/group-anagrams/}
export { groupAnagrams };

// SOLUTION:
//
// Each anagram can be rearranged into canonical form by sorting the letters.  Then simply map the canonical form to
// each anagram.
//
// COMPLEXITY:
//
// Runs in O(n * m * log(m)) time, where n is the number of strings and m is the length of the longest string.  This
// is because we have to sort each string's characters in O(m * log(m)), and there are n strings.
function groupAnagrams(texts: string[]): string[][] {
  type Canonical = string;
  type Anagram = string;
  const map = new Map<Canonical, Anagram[]>();

  // Group all the anagrams together by their "canonical" form (aka a sorted version of the string).
  for (let i = 0; i < texts.length; i++) {
    const text = texts[i];
    const canonical = text.split('').sort().join('');
    const anagrams = map.get(canonical) ?? [];

    anagrams.push(text);
    map.set(canonical, anagrams);
  }

  // The map.values() function gives you an iterable of string[], so we need to convert it to an array of string[]
  // instead.
  return Array.from(map.values());
}
