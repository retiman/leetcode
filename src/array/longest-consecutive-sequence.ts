// DIFFICULTY: Medium
//
// Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.
//
// You must write an algorithm that runs in O(n) time.
//
// See {@link https://leetcode.com/problems/longest-consecutive-sequence/}
export { longestConsecutive };

// SOLUTION:
//
// The problem is phrased in a very confusing way.  The sequence doesn't ACTUALLY need to be consecutive within the
// array; it only needs to be consecutive if you plucked the sequence out of the array and sorted it.
//
// For example, [1, 500, 2, 3, 4] has consecutive elements [1, 2, 3, 4].  The fact that 500 appears in the middle
// is okay; the number 500 begins its own consecutive sequence of length 1.  The other elements [1, 2, 3, 4] create
// a sequence of length 4.
//
// Conceptually, we'll do this by throwing all the array elements into a set.  Then, for each element `x`, we can figure
// out if it's part of a sequence by repeatedly checking its predecessor `x - 1` in the set.
//
// COMPLEXITY:
//
// Runs in O(n) time.  It may appear that the inner loop runs multiple times, but each element in the array is only
// processed once; the inner loop will skip `x - 1` if it was already part of some other sequence from a previous
// iteration.
function longestConsecutive(xs: number[]) {
  let longest = 0;

  // Use a set to keep track of all elements in the array; we'll reference it to find out if a predecessor to an
  // element exists as we iterate through the array.
  const set = new Set(xs);

  for (let i = 0; i < xs.length; i++) {
    // Let us consider if element x is part of some sequence.  We can do this by considering its predecessor, x - 1.
    let x = xs[i];

    // If x - 1 is not in the set, x must begin some new sequence.  Let's find out how long it is by incrementing x
    // until we can't find any more consecutive elements.
    if (!set.has(x - 1)) {
      let length = 1;
      x++;

      // Continue to increment x and look for consecutive elements in the set.
      while (set.has(x)) {
        length++;
        x++;
      }

      // Once we run out of elements, we have determined the length of the sequence between at x.  We'll compare it to
      // the longest sequence we've found so far.
      longest = Math.max(longest, length);
    } else {
      // If x - 1 is in the set, we know it is part of some sequence.  However, this sequence must've already been
      // found by the inner loop above, so we can skip this element.
    }
  }

  return longest;
}
