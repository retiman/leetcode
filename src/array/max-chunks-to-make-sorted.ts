// DIFFICULTY: MEDIUM
//
// You are given an integer array arr of length n that represents a permutation of the integers in the range [0, n - 1].
//
// We split arr into some number of chunks (i.e., partitions), and individually sort each chunk. After concatenating
// them, the result should equal the sorted array.
//
// Return the largest number of chunks we can make to sort the array.
//
// See {@link https://leetcode.com/problems/max-chunks-to-make-sorted/}
export { maxChunksToSorted };

// SOLUTION:
//
// Because we know the elements in the array are a permutation of numbers less than n, we can use a greedy approach
// by keeping track of the max element seen so far.
//
// Unlike the other problem, this does not require using a stack.
function maxChunksToSorted(xs: number[]) {
  let chunks = 0;
  let max = 0;

  for (let i = 0; i < xs.length; i++) {
    max = Math.max(max, xs[i]);

    // Because all elements are a permutation of the numbers from 0 to xs.length - 1, if we encounter i === max,
    // this means that for sure all elements < i can be sorted to form a chunk.
    //
    // Again, because all elements to the right of i are going to be greater than max (and i), then we can be sure
    // that the next element begins a new chunk.
    if (i === max) {
      chunks++;
    }
  }

  return chunks;
}
