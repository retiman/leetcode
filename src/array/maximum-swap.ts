// DIFFICULTY: MEDIUM
//
// You are given an integer num. You can swap two digits at most once to get the maximum valued number.
//
// Return the maximum valued number you can get.
//
// See {@link https://leetcode.com/problems/maximum-swap/}
export { maximumSwap };

// SOLUTION:
//
// The simple idea of swapping the largest number with the first digit doesn't always work.  It works a lot of the time,
// but a couple of cases mess up this algorithm:
//
// - 98368 would stay 98368; we need to swap the 3 and 8 to get 98863.
// - 1993 could become 9193, but we need to swap the 1 and other 9 to get 9913.
//
// The way to think about this is:
//
// 1. Find the FIRST digit that can be made bigger (larger digits appear later).
// 2. Swap it with the LAST occurrence of the largest digit.
//
// COMPLEXITY:
//
// The time complexity is O(n) where n is the number of digits.  We iterate through the digits once to create our array
// and map.  We iterate through the digits a second time to find the swap point.  There is a nested loop, but the inner
// loop is bounded by the 10 digits, so it doesn't cause O(n^2) time complexity.
//
// The space complexity is O(n) because we store an array and map of the digits.
function maximumSwap(num: number): number {
  // First convert the number to an array of digits.
  const digits = num.toString().split('').map(Number);

  // Create a map of the last seen index of each digit.  Use this to find out the last occurrence of any digit.
  type Digit = number;
  type LastSeen = number;
  const map = new Map<Digit, LastSeen>();
  for (let i = 0; i < digits.length; i++) {
    const digit = digits[i];
    map.set(digit, i);
  }

  // Find the first digit that can be made bigger by iterating through the digits list.
  for (let i = 0; i < digits.length; i++) {
    const smaller = digits[i];

    // Find a larger digit to swap the smaller digit with.  To do so, start at 9 and iterate downwards until we find a
    // digit that is larger, and appears later in the number.
    for (let larger = 9; larger >= 0; larger--) {
      // Skip the digit if it's not larger.
      if (larger <= smaller) {
        break;
      }

      // Skip the digit if it doesn't appear at all.
      if (!map.has(larger)) {
        continue;
      }

      // Skip the digit if it appears before the smaller digit.
      const j = map.get(larger)!;
      if (j <= i) {
        continue;
      }

      // Success!  We have found the largest digit that occurs as late as possible!  Swap the digits and return the
      // number.
      [digits[i], digits[j]] = [digits[j], digits[i]];
      return Number(digits.join(''));
    }
  }

  // We made no swaps, so just return the number as is.
  return num;
}
