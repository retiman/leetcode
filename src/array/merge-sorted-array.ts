// DIFFICULTY: EASY
//
// You are given two integer arrays nums1 and nums2, sorted in non-decreasing order, and two integers m and n,
// representing the number of elements in nums1 and nums2 respectively.
//
// Merge nums1 and nums2 into a single array sorted in non-decreasing order.
//
// The final sorted array should not be returned by the function, but instead be stored inside the array nums1. To
// accommodate this, nums1 has a length of m + n, where the first m elements denote the elements that should be merged,
// and the last n elements are set to 0 and should be ignored. nums2 has a length of n.
//
// See {@link https://leetcode.com/problems/merge-sorted-array/}
export { merge };

// SOLUTION:
//
// This would be VERY easy, if you could use extra memory.  To merge without using extra memory, we do need to merge
// into the bigger array, nums1.
//
// You could go about this in two different ways:
//
// 1. Start from the logical beginning of both arrays and swap elements as needed to maintain the correct order.
// 2. Start from the logical end of both arrays and add largest elements to the end of nums1.
//
// Going with option 1 is more natural, but it's more error prone and more work.  There may be elements at the end of
// nums1 that need to be shifted to the right to make room.
//
// Going with option 2 is less natural, but it ensures you have enough space.  Since nums1 has size m + n, and nums2 has
// size n, you can reliably fit all of nums2 into nums1 without any collisions or shifting.  Additionally, if you use
// up all the elements in nums2, you don't need to do anything with the remaining elements of nums1 since they are
// already sorted.
//
// Pro tip: when asked to do something in place, and you have extra space, consider doing backwards iteration instead of
// forwards to avoid collisions and overwriting elements.  Secondly, if you started with forwards iteration and realize
// you might have to shift elements, consider a backwards iteration approach to see if it might work better.
//
// The problem doesn't state this, but we assume using no extra memory either.
function merge(nums1: number[], m: number, nums2: number[], n: number): void {
  let i = m - 1;
  let j = n - 1;
  let last = m + n - 1;

  // Send the largest elements from both arrays to the end of the array nums1.
  while (i >= 0 && j >= 0) {
    const a = nums1[i];
    const b = nums2[j];

    if (a < b) {
      nums1[last] = nums2[j];
      j--;
      last--;
    } else {
      nums1[last] = nums1[i];
      i--;
      last--;
    }
  }

  // Now we have consumed all elements in one of the arrays.
  //
  // If there are remaining elements in nums2, we should add them to the array nums1.
  while (j >= 0) {
    nums1[last] = nums2[j];
    j--;
    last--;
  }

  // On the other hand, if there are remaining elements in nums2, there's nothing to be done because that slice of the
  // array is already sorted.
}
