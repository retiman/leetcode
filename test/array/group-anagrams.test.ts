import { groupAnagrams } from "../../src/array/group-anagrams";

describe('group anagrams', () => {
  test('group anagrams - test case 1', async () => {
    expect(groupAnagrams(['eat', 'tea', 'tan', 'ate', 'nat', 'bat'])).toMatchSnapshot();
  });
});
