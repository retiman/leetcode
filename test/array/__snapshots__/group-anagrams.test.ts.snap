// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`group anagrams group anagrams - test case 1 1`] = `
[
  [
    "eat",
    "tea",
    "ate",
  ],
  [
    "tan",
    "nat",
  ],
  [
    "bat",
  ],
]
`;
